cmd/rule-evaluator/main.go

```diff
@@ -959,17 +969,60 @@ func newRuleEvaluator(
 	})
 
 	evaluator := ruleEvaluator{
+		ctx:             ctx,
+		logger:          logger,
+		version:         version,
+		appendable:      appendable,
+		notifierManager: notifierManager,
+		rulesMetrics:    rulesMetrics,
+
 		rulesManager:      rulesManager,
 		queryFunc:         queryFunc,
 		lastEvaluatorOpts: evaluatorOpts,
+		rerunManager:      false,
 	}
 
 	return &evaluator, nil
 }
 
 func (e *ruleEvaluator) ApplyConfig(cfg *config.Config, evaluatorOpts *evaluatorOptions) error {
 	if evaluatorOpts != nil && !reflect.DeepEqual(evaluatorOpts, e.lastEvaluatorOpts) {
-		return fmt.Errorf("rule-config can only be changed at startup time")
+		e.lastEvaluatorOpts = evaluatorOpts
+
+		v1api, err := newAPI(e.ctx, evaluatorOpts, e.version, func(rt http.RoundTripper) http.RoundTripper {
+			return makeInstrumentedRoundTripper(rt)
+		})
+		if err != nil {
+			return fmt.Errorf("query client: %w", err)
+		}
+		queryFunc := newQueryFunc(e.logger, v1api)
+
+		rulesManager := rules.NewManager(&rules.ManagerOptions{
+			ExternalURL: getExternalURL(evaluatorOpts.GeneratorURL, evaluatorOpts.ProjectID),
+			QueryFunc:   queryFunc,
+			Context:     e.ctx,
+			Appendable:  e.appendable,
+			Queryable: &queryStorage{
+				api: v1api,
+			},
+			Logger:     e.logger,
+			NotifyFunc: sendAlerts(e.notifierManager, evaluatorOpts.ProjectID, evaluatorOpts.GeneratorURL),
+			Metrics:    e.rulesMetrics,
+		})
+
+		// Set new rule-manager and flag before stopping, so we can rerun with the new one.
+		e.mtx.Lock()
+		oldRuleManager := e.rulesManager
+		e.rulesManager = rulesManager
+		e.rerunManager = true
+		oldRuleManager.Stop()
+		e.queryFunc = queryFunc
```

Comment:
I believe the user of `queryFunc` should also be behind the mutex in this scenario.

```
func (e *ruleEvaluator) Query(ctx context.Context, q string, t time.Time) (promql.Vector, error) {
	return e.queryFunc(ctx, q, t)
}
```